### ISSUE TITLE
Consider maybe resyncing with upstream?

### ISSUE THREAD

**Issue by mitar** [sudomesh/tunneldigger#8]
There are many changes done upstream, contributed by various communities. I suggest this repo is synced with them and changes pushed upstream to the main repo.

**Comment by mitar** [comment]
See also #3.

**Comment by jhpoelen** [comment]
@mitar that makes total sense to me . . . right now, we're attempting to the exitnode (see https://github.com/sudomesh/bugs) up and running, so we can reach existing nodes for patching . 

As far as I know, the home node have an older tunneldigger client that won't work with the new broker. Did you try running the older client against the latest tunneldigger broker?

**Comment by mitar** [comment]
Do you know which exactly version of the older code they are running?

I know that we had some tests making sure various versions still work with each other. I will cc @RalfJung here, who was the last to work on this and make he can help a bit to tell us which versions he knows to still work together.

**Comment by jhpoelen** [comment]
As far as I can tell the last sync was https://github.com/sudomesh/tunneldigger/commit/cfe219e2a8c4af5be811f583dc74da1282bb51ad .

Would be great to know more about client/broker compatibility .

**Comment by mitar** [comment]
See [changelog](https://github.com/wlanslovenija/tunneldigger/blob/master/HISTORY.rst) for some information.

**Comment by jhpoelen** [comment]
As far as I can tell, the exit node is able to create only a single tunnel. All attempts to create additional tunnels fail. The exit node is running

```
$ uname -a
Linux exit 3.16.0-4-amd64 #1 SMP Debian 3.16.51-3 (2017-12-13) x86_64 GNU/Linux
```

Running the same version of tunneldigger on ubuntu 16.04.3 on Digital Ocean droplets, does not seem to have this limitation. These droplets are running:

```
# uname -a
Linux dig-that-tunnel 4.4.0-112-generic #135-Ubuntu SMP Fri Jan 19 11:48:36 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by mitar** [comment]
This does not seems to be consistent. you are saying that old kernels can have only one tunnel, and newer multiple, but https://github.com/wlanslovenija/tunneldigger/issues/57 is the opposite.

**Comment by jhpoelen** [comment]
Yes, this would assume that both ubuntu and debian would have included the path in respective kernel versions. My current theory is that patch https://patchwork.kernel.org/patch/9843481/ is not yet included in 16.04.3, whereas is was back-ported to debian 3.16.51-3  . 

Here's the related patch https://patchwork.kernel.org/patch/9843481/ . Trying to investigate more and curious to hear thoughts.

**Comment by jhpoelen** [comment]
related to https://github.com/sudomesh/bugs/issues/8

**Comment by mitar** [comment]
Yea, I really think you should upgrade Tunneldigger. :-)

**Comment by jhpoelen** [comment]
I agree that we should upgrade tunnel digger. 

And we have to figure out a way to update the deployed home nodes to newer tunnel digger client to avoid having to knock on doors and get physical access to 50+ deployed nodes. 

It seems to me that downgrading to a kernel version that does not include the fix would be one way to re-establish tunnels so that nodes can be patched.

@mitar curious to hear your thoughts on this pragmatic concern

**Comment by jhpoelen** [comment]
An alternative would be to remotely patch nodes one at a time.

**Comment by mitar** [comment]
The session issue is just on the server. Not clients. We have not had to upgrade any client for them to keep working.

**Comment by mitar** [comment]
But you might want to have an old and new server around. So in our case we have multiple VPN servers so we can roll out updates one by one.

**Comment by gobengo** [comment]
@jhpoelen What do you think about having exitnode keep running old tunneldigger on existing port and now *also* newest tunneldigger with distinct session IDs on a new port? The former will be deprecated and turned off once no more clients use it (or in 2 years or something reasonable).

**Comment by RalfJung** [comment]
In terms of compatibility, old and new client and broker should be compatible in all combinations.

*However*, to actually be able to run a kernel with the bugfix (i.e., a kernel enforcing unique session IDs), both the broker and all the clients connecting to it must be updated.

> The session issue is just on the server. Not clients. We have not had to upgrade any client for them to keep working.

I don't know how this should be possible. The old client hard-codes the session ID of *both* ends of the tunnel to be 1, it is impossible to have two such clients establish an L2TP session with a single host running a recent kernel. If an old client connects to a new broker, that will work, but it won't actually use the fix -- so new kernels can only have one such client connected at any given time.

**Comment by jhpoelen** [comment]
We got the old tunneldigger working for now by downgrading our debian kernel. It appears that Ubuntu 16.04.3 LTS hasn't yet incorporated the l2tp patch, to old tunnelbroker still works on that.

@gobengo @mitar thanks for sharing ideas for upgrade path - I've created an example configuration for tunnel broker clients on home nodes https://github.com/sudomesh/bugs/issues/17 . (@gobengo want to give this one a shot?)

@RalfJung thanks for chiming in - you confirmed the behavior I was seeing. So, yes, the old client works with the new brokers, but you'd have to have as many brokers as clients for this to work . . . and the brokers need to mesh with each other in order to creates routes across the old client. Hoping we can upgrade the home node clients remotely and gradually move them over to new brokers.

**Comment by mitar** [comment]
I would propose that you run one old and one new version of Tunneldigger, but run old kernel. Then both old and new can connect. And this gives you time to upgrade all home devices.

**Comment by Juul** [comment]
Does meltdown mitigation require both host and vm kernel to be patched?

**Comment by mitar** [comment]
@Juul, I think there are multiple things going on. But the main thing is that in newer kernels session IDs for tunnels have to be unique. New Tunneldigger does that, old does not. So if you want to keep running old (for old clients which use fixed session ID) you can simply run an older kernel. There is no patch involved. And this is just something to run wherever you run your tunnels, probably inside VM.

**Comment by RalfJung** [comment]
@jhpoelen 
> So, yes, the old client works with the new brokers, but you'd have to have as many brokers as clients for this to work 

Just to be clear, this is only if you use the new *kernel*. So the upgrade path is:
* Keep the old kernels but upgrade all brokers, and the clients. Because of mutual compatibility, you can update brokers and clients in any order.
* Once all brokers and clients are updated, update the kernels.

What we are doing is a slight refinement, where we already have new kernels running on 2 of the 4 brokers; the few remaining old clients will use the remaining 2 brokers that still use an old kernel.
